Move to Shared - SqlSer.cs

Kaur-Parminder opened the pull request:
This is part of Code merge for issue #1261. I have created .common.cs and moved the uncommon code to private methods.

Wraith2 commented on src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/Server/SqlSer.cs:
```diff
@@ -151,72 +30,19 @@ internal static SqlUserDefinedTypeAttribute GetUdtAttribute(Type t)
             ((IBinarySerialize)o).Write(w);
         }
 
-        // Prevent inlining so that reflection calls are not moved
-        // to a caller that may be in a different assembly that may
-        // have a different grant set.
-        [MethodImpl(MethodImplOptions.NoInlining)]
```
This attribute and it's explanation seem important to keep.

Kaur-Parminder replied:
@Wraith2 This method, attribute and comment has been moved to SqlSer.Common.cs

Kaur-Parminder replied:
@Wraith2  This methods, attribute and comment has been moved to SqlSer.Common.cs

Wraith2 replied:
Ok. Diff comprehension failure on my part then.

Kaur-Parminder replied:
Understood.

Kaur-Parminder commented:
> There's not that much code remaining in the framework specific files. Why don't we just move it to the common file?

@johnnypham I just removed the netfx and netcore classes and updated SqlSer.cs in common project with the three functions. 
1. Both Serialize and DeSerialize functions uses If defs, added return statements in if block.
2 .GetUdtAttribute function in netfx used Reflection to throw same exception so just kept the .NetCore one.

DavoudEshtehari reviewed (commented):
The `SerializationHelperSql9` could be sealed: `internal sealed class SerializationHelperSql9`
Also, I see some minor possible improvements like `unnecessary assignment` and `redundant member`. You can find them in the `error list` window.

DavoudEshtehari commented on src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Server/SqlSer.cs:
```diff
@@ -209,7 +207,16 @@ internal BinarySerializeSerializer(Type t) : base(t)
             BinaryWriter w = new BinaryWriter(s);
+
+#if NETFRAMEWORK
+            if (o is Microsoft.SqlServer.Server.IBinarySerialize)
+            {
+                ((SqlServer.Server.IBinarySerialize)o).Write(w);
+                return;
+            }
```
```suggestion
            if (o is Microsoft.SqlServer.Server.IBinarySerialize bs)
            {
                bs.Write(w);
                return;
            }
```

DavoudEshtehari commented on src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Server/SqlSer.cs:
```diff
@@ -220,8 +227,17 @@ public override object Deserialize(Stream s)
-            ((IBinarySerialize)instance).Read(r);
+
+#if NETFRAMEWORK
+            if (instance is Microsoft.SqlServer.Server.IBinarySerialize)
+            {
+                ((SqlServer.Server.IBinarySerialize)instance).Read(r);
+                return instance;
+            }
```
```suggestion
            if (instance is Microsoft.SqlServer.Server.IBinarySerialize bs)
            {
                bs.Read(r);
                return instance;
            }
```

DavoudEshtehari reviewed (commented):
I'd suggest using `Dictionary` instead of `Hashtable` to avoid boxing and unboxing for [s_types2Serializers](https://github.com/dotnet/SqlClient/blob/04593ff73078c40e3bf83d231b44be935455763b/src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Server/SqlSer.cs#L54):
- `private static ConcurrentDictionary<Type, Serializer> s_types2Serializers;`

Wraith2 commented on src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Server/SqlSer.cs:
```diff
@@ -50,13 +52,14 @@ internal static void Serialize(Stream s, object instance)
             if (s_types2Serializers == null)
-                s_types2Serializers = new Hashtable();
+            {
+                s_types2Serializers = new();
```
The rule used in /runtime is that use of `var` or `new()` is only permitted if the type is duplicated in the same statement. So either `var s_types2Serializers = new ConcurrentDictionary<Type, Serializer>()` or `ConcurrentDictionary<Type, Serializer> s_types2Serializers = new ()` would be allowed (with var preferred I think) but not the form used here.

If you rebase to master the style rules were updated in a recent PR to remove the new() fixer suggestion in the main library projects.

DavoudEshtehari commented on src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Server/SqlSer.cs:
```diff
@@ -4,14 +4,15 @@
 
 using System;
 using System.Collections;
```
```suggestion
```

DavoudEshtehari commented on src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Server/SqlSer.cs:
```diff
@@ -50,16 +52,21 @@ internal static void Serialize(Stream s, object instance)
 
-            Serializer s = (Serializer)s_types2Serializers[t];
-            if (s == null)
+            if (s_types2Serializers.ContainsKey(t))
```
This `if-else` block could be replaced with the following block:
```
if(!s_types2Serializers.TryGetValue(t, out s))
{
    s = GetNewSerializer(t);
    s_types2Serializers[t] = s;
}
```

Kaur-Parminder replied:
TryGetValue will return Default value if there is no value for the key, condition will never be false and compiler will not enter if block. Although, I think I should keep the check if(s==null) after contains key check because we don't check  null value before entry to dictionary.

DavoudEshtehari commented on src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Server/SqlSer.cs:
```diff
@@ -50,16 +52,21 @@ internal static void Serialize(Stream s, object instance)
 
         private static Serializer GetSerializer(Type t)
         {
+            Serializer s = null;
```
NIT: Unnecessary assignment.
```suggestion
            Serializer s;
```
